File: javascript/092.ReverseLinkedListII.js
// https://leetcode.com/problems/reverse-linked-list-ii/

// https://leetcode.com/problems/reverse-linked-list-ii/

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} m
 * @param {number} n
 * @return {ListNode}
 */
var reverseBetween = function (head, m, n) {
    if (m === n || !head) {
        return head
    }
    let dummy = new ListNode(null)
    dummy.next = head
    let pre = dummy
    let i = 1
    while (i < m) {
        pre = pre.next
        i++
    }
    let current = pre.next
    let then = current.next
    while (i < n) {
        current.next = then.next
        then.next = pre.next
        pre.next = then
        then = current.next
        i++
    }
    return dummy.next
};

File: python/092.ReverseLinkedListII.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def reverseBetween(self, head: ListNode, m: int, n: int) -> ListNode:
        if m == n:
          return head
        newHead = ListNode(None)
        newHead.next = head
        pre = newHead
        for i in range(m - 1):
          pre = pre.next

        start = pre.next
        then = start.next
        for i in range(n - m):
          start.next = then.next
          then.next = pre.next
          pre.next = then
          then = start.next

        return newHead.next
